### DISCUSSION TITLE
Recommended way to use external force constants with ANPHON

### DISCUSSION THREAD

**Discussion by ejmeitz** [ttadano/alamode#174] [Q&A]
I would like to use the ANPHON part of ALAMODE to calculate thermal conductivity, DOS etc. for various approaches of calculating IFCs. One of these approaches is of course the optimization approach included in ALMODE but I would also like to plug in the 0K force constants (from finite displacement) and just in general force constants from other packages (e.g. TDEP).

Reading through the documentation the IFC format (.fcs) does not look too complex, but the XML format that is needed by ANPHON is not something I feel comfortable generating. My general question is are there ways for me to convert the .fcs to the .xml format needed by ANPHON or is there a way I could just give a big list of force constants for a supercell (not symmetry reduced) and have ANPHON understand that input.

Thanks!

**Comment by ttadano** [comment]
Thank you for the suggestion. I am aware that the format of FCS XML is too complicated, so I will introduce a new format (HDF5) in the future major release of ALAMODE.

For the moment, please use the python module at https://github.com/ttadano/ALM.
You can install the module as follows:

```
git clone https://github.com/ttadano/ALM
cd ALM
git checkout develop
cd python
python setup.py install
```

After installing the module, please have a look at `example/write_fcsxml.py` and try
```
python write_fcsxml.py
```
This is a minimal example showing how to save the force constants (obtained by other method) as the FCS XML format readable by ANPHON.

The usage is simple. You just need to give the structural parameters, force constant values and corresponding indices as follows:

```
obj = Fcsxml(lavec, xcoord, kd)
obj.set_force_constants(fc_values, elem_indices)
obj.set_force_constants(fc3_values, elem3_indices)
obj.write(filename="test.xml")
```

For more details about the `Fcsxml` module, please issue `pydoc alm.fcsxml`.

**Reply by ejmeitz** [reply]
Thanks!

**Reply by ejmeitz** [reply]
Any reason you can think of my spglib install just made shared libraries? The `setup.py` script seems to be looking for the static library `.a` version which was not installed by provided command:

`conda install -c conda-forge numpy scipy h5py compilers "libblas=*=*mkl" spglib boost eigen cmake ipython mkl-include`

```
libsymspg.so
libsymspg.so.2
libsymspg.so.2.3.1
```

**Reply by ttadano** [reply]
Hi, 

I've updated `setup.py` script of the ALM repo. Please `git pull`, cleanup the previous build, and run `python -m pip install .` following the documentation https://alm.readthedocs.io/en/latest/compile-with-conda-packages.html.

**Reply by ejmeitz** [reply]
Build succeeded, Thank you!

**Comment by ejmeitz** [comment]
I was digging into the example, but some of the intermediate output is not what I expected.

From my understanding, the `xcoord` in the Fcsxml example is from a 2x2x2 unitcell of silicon which has a 2 atom primitive cell or an 8 atom conventional cell. Specifically, I am confused by the indices output by `fc_values, elem_indices = alm.get_fc(1)`. The default mode is `"origin"` so the first element in the index "corresponds to an atom in the primitive cell at the origin". In this example what does that mean? If I actually print out the unique values for the first index I get (0,1,2,24,25,26) which correspond to the following atoms:

```
[0.0, 0.0, 0.0]
[0.0, 0.0, 0.5]
[0.0, 0.25, 0.25]
[0.375, 0.125, 0.375]
[0.375, 0.125, 0.875]
[0.375, 0.375, 0.125]
```

which don't look like the atoms in the conventional or primitive cell to me. I guess I'm just confused about which force constants are included in the output of `get_fc`. Any help would be appreciated!

**Reply by ttadano** [reply]
Hi,

The indices `(0,1,2,24,25,26)` are flattened index of atomic index and coordinate (x, y, z).
These numbers should be interpret as follows:

(0, 1, 2) --> (1x, 1y, 1z) where 1 is the label of the first atom in the supercell
(24, 25, 26) --> (9x, 9y, 9z)

The 1st and 9th atoms in the supercell form a primitive cell.

**Comment by ejmeitz** [comment]
I got things working, but just a heads up I think the units are outdated for that example.

The default units for Fcxsml are 'angstrom' (see [here](https://github.com/ttadano/ALM/blob/6a70b30fcd64c56d581d828f7ec754dd604c723e/python/alm/fcsxml.py#L137)) but everything in that example is Bohr as far as I can tell. It's not a big deal, but if you are just copying that example code it gives weird results.
